### ISSUE TITLE
postgre function IN OUT parameters ordering not working properly

### ISSUE THREAD

**Issue by nikwan** [impossibl/pgjdbc-ng#207]
IN OUT parameters not working properly if having OUT c timestamp without time zone type giving exception.
IN OUT ordering after executing method from java not working properly see below example

```
 call.registerOutParameter(2, java.sql.Types.VARCHAR);
            call.registerOutParameter(3, java.sql.Types.VARCHAR);
            call.registerOutParameter(4, java.sql.Types.BIGINT); // VARCHAR to NUMERIC
            call.registerOutParameter(5, java.sql.Types.VARCHAR);
            call.registerOutParameter(6, java.sql.Types.VARCHAR);
            call.registerOutParameter(7, java.sql.Types.VARCHAR);
            call.registerOutParameter(8, java.sql.Types.VARCHAR);
            call.registerOutParameter(9, java.sql.Types.TIMESTAMP);
            call.registerOutParameter(10, java.sql.Types.BIGINT);
            call.registerOutParameter(11, java.sql.Types.VARCHAR);
            call.registerOutParameter(12, java.sql.Types.VARCHAR);
            call.registerOutParameter(13, java.sql.Types.VARCHAR);
            call.registerOutParameter(14, java.sql.Types.VARCHAR);
            call.registerOutParameter(15, java.sql.Types.BIGINT);
            call.registerOutParameter(16, java.sql.Types.VARCHAR);
            call.registerOutParameter(17, java.sql.Types.VARCHAR);
            call.registerOutParameter(18, java.sql.Types.BIGINT);               

            boolean b = call.execute();
            System.out.println(call.getString(2));
            System.out.println(call.getString(3));
            System.out.println(call.getLong(4));
            System.out.println(call.getString(5));
            System.out.println(call.getString(6));
            System.out.println(call.getTimestamp(9));
```

  java.sql.SQLException: Coercion from 'java.lang.Long' to 'java.sql.Timestamp' is not supported (0)
    at com.impossibl.postgres.jdbc.SQLTypeUtils.createCoercionException(SQLTypeUtils.java:1187)
    at com.impossibl.postgres.jdbc.SQLTypeUtils.coerceToTimestamp(SQLTypeUtils.java:641)
    at com.impossibl.postgres.jdbc.PGCallableStatement.getTimestamp(PGCallableStatement.java:513)
    at com.impossibl.postgres.jdbc.PGCallableStatement.getTimestamp(PGCallableStatement.java:486)
    at com.main.T.oracleToPosgresUDTTestCall(T.java:410)
    at com.main.T.main(T.java:33)
  java.sql.SQLException: Coercion from 'java.lang.Long' to 'java.sql.Timestamp' is not supported (0)

**Comment by jesperpedersen** [comment]
Would you be interested in submitting a test case for src/test/java/com/impossibl/postgres/jdbc/TimestampTest.java ?

The required updates to SQLTypeUtils.java should be easy too.

**Comment by nikwan** [comment]
Thanks for reply.
I tested on my local environment.
Below is the postgres9.4 db function

```
-- Function: n_test_in_out_index(text)

-- DROP FUNCTION fn_test_in_out_index(text);

CREATE OR REPLACE FUNCTION fn_test_in_out_index(
    IN txt text,
    OUT a text,
    OUT b text,
    OUT c timestamp without time zone,
    OUT d text,
    OUT e text,
    OUT f text,
    OUT g text,
    OUT h text,
    OUT i text,
    OUT j text,
    OUT k text,
    OUT l text,
    OUT m text,
    OUT n text,
    OUT o text,
    OUT p text,
    OUT q bigint)
  RETURNS record AS
$BODY$
DECLARE 
BEGIN
RAISE INFO '@@ inside fn_test:(%)',111;
a := 'test'; 
b := 'test'; 
c := LOCALTIMESTAMP;
d := 'hi nikhil';
q := 111;
RAISE INFO 'a a:(%)',a;
RAISE INFO 'b b:(%)',b;
RAISE INFO 'LOCALTIMESTAMP c:(%)',c;
RAISE INFO 'd d:(%)',d;
  --  RETURN address;
END;
$BODY$
  LANGUAGE plpgsql VOLATILE
  COST 100;
ALTER FUNCTION fn_test_in_out_index(text)
  OWNER TO postgres;
```

below is the main function to test.
Ordering of IN OUT parameters are not working as expected.
Please remove comment from the below function test call to test first parameter which raise you exception.

```
public static void oracleToPosgresINOUTParamTest() {
        System.out.println("@@@ inside oracleToPosgresINOUTParamTest...");
        Connection c = null;
        try {
            Class.forName("com.impossibl.postgres.jdbc.PGDriver");
            c = DriverManager.getConnection(
                    "jdbc:pgsql://localhost:5433/orapg", "postgres", "root");
            System.out.println("driver version:" + c.getMetaData().getDriverVersion());

            // you can mad your UDT to pojo here Great pgjdbc-ng !!!
            /*Map<String, Class<?>> m = c.getTypeMap();
            m.put("typ_address", Address.class);
            m.put("typ_test", AddressTypeTest.class);
            c.setTypeMap(m);*/

            // Procedure call
            CallableStatement cstmt = c
                    .prepareCall("{call fn_test_in_out_index(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)}");

            cstmt.setString(1, "test");

            cstmt.registerOutParameter(2, Types.VARCHAR);
            cstmt.registerOutParameter(3, Types.VARCHAR);
            cstmt.registerOutParameter(4, Types.TIMESTAMP);
            cstmt.registerOutParameter(5, Types.VARCHAR);
            cstmt.registerOutParameter(6, Types.VARCHAR);
            cstmt.registerOutParameter(7, Types.VARCHAR);
            cstmt.registerOutParameter(8, Types.VARCHAR);
            cstmt.registerOutParameter(9, Types.VARCHAR);
            cstmt.registerOutParameter(10, Types.VARCHAR);
            cstmt.registerOutParameter(11, Types.VARCHAR);
            cstmt.registerOutParameter(12, Types.VARCHAR);
            cstmt.registerOutParameter(13, Types.VARCHAR);
            cstmt.registerOutParameter(14, Types.VARCHAR);
            cstmt.registerOutParameter(15, Types.VARCHAR);
            cstmt.registerOutParameter(16, Types.VARCHAR);
            cstmt.registerOutParameter(17, Types.VARCHAR);
            cstmt.registerOutParameter(18, Types.BIGINT);

            boolean b = cstmt.execute();

            //////////////////////////////////////////////////////////////////////
            //String s1 = cstmt.getString(1); //remove comment to test 
            // output string index
            // registerOutParameter with 2 but
            // returning with 1
            /* if get with index 1 throws SQLException 
             * java.sql.SQLException: parameter not available
             * at com.impossibl.postgres.jdbc.PGCallableStatement.mapToOutParameterIndex(PGCallableStatement.java:226)
             * at com.impossibl.postgres.jdbc.PGCallableStatement.getString(PGCallableStatement.java:373)
             */
            /////////////////////////////////////////////////////////////////////

            String s = cstmt.getString(2); // output string index
                                            // registerOutParameter with 2 but
                                            // returning with 1
            System.out.println("2:" + s);

            Timestamp t = cstmt.getTimestamp(3); // output TIMESTAMP index
                                                    // registerOutParameter with
                                                    // 4 but returning with 3
            System.out.println("4:" + t);

            System.out.println("18:" + cstmt.getLong(18)); // output BIGINT
                                                            // index
                                                            // registerOutParameter
                                                            // with 18 but
                                                            // returning
                                                            // correctly 18


            cstmt.close();

        } catch (Exception e) {
            e.printStackTrace();

        }
        System.out.println("test finished...");
    }
```

**Comment by kdubb** [comment]
Fixed in f8f656d369a2b69ed97702def6475c487ea2bcb9. Added a version of this function to a test in `CallableStatementTest`
